lib/module-entry.js

```diff
@@ -0,0 +1,227 @@
+ */
+ModuleEntry.prototype._gatherDependencies = function(dependencies) {
+  var h = heimdall.start('hashForDep:gatherDependencies');
+
+  try {
+    var moduleEntry = this;
+    var rootDir = moduleEntry.rootDir;
+
+    if (dependencies[rootDir] !== undefined) {
+      // already gathered moduleEntry dependency, so exit early to avoid doing
+      // duplicate work
+      return;
+    }
+
+    dependencies[rootDir] = moduleEntry;
+
+    Object.keys(moduleEntry._dependencies).forEach(function(dep) {
+      moduleEntry._dependencies[dep]._gatherDependencies(dependencies);
+    });
+  } finally {
+    h.stop();
+  }
+
+  return dependencies;
+};
+
+function genRealPathKey(realPath) {
+  return cacheKey('hashed:' + realPath, '');
+}
+
+/*
+ * Compute and return the ModuleEntry for a given name/dir pair. This is done
+ * recursively so a whole tree can be computed all at once, independent of the
+ * hashing functions. We also compute the 'sourceHash' value (i.e. the hash of
+ * just the package source files, excluding the node_modules dependencies) at
+ * the same time to make it easier to establish the 'hash' value (includes
+ * dependency sourceHash values) later.
+ */
+ModuleEntry.locate = function(caches, name, dir, hashTreeFn) {
+  var h = heimdall.start('hashForDep:computeModuleEntry');
+  var Constructor = this;
+
+  try {
+    var nameDirKey = genNameDirKey(name, dir);
+    var realPathKey;
+
+    // It's possible that for a given name/dir pair, there is no package.
+    // We record a null entry in CACHES.PATH anyway so we don't need to
+    // redo that search each time.
+    if (caches.PATH.has(nameDirKey)) {
+      realPathKey = caches.PATH.get(nameDirKey);
+      if (realPathKey !== null) {
+        return caches.MODULE_ENTRY.get(realPathKey);
+      } else {
+        return null;
+      }
+    }
+
+    // We don't have a caches.PATH entry. See if we can find a
+    // real package.json path. If we can't, the reference is
+    // invalid, so we'll note that in caches.PATH. If we do find
+    // one, see if we already have the caches.MODULE_ENTRY entry for it,
+    // then create it if not.
+    var realPath = resolvePackagePath(caches.REAL_FILE_PATH, name, dir);
+
+    if (realPath === null) {
+      caches.PATH.set(nameDirKey, null);
+      return null;
+    }
+
+    // We have a path to a file that supposedly is package.json.
+    // We need to be sure that either we already have a CACHES.MODULE_ENTRY
+    // entry (in which case we can just create a CACHES.PATH entry to
+    // point to it) or that we can read the package.json file, at which
+    // point we can create a CACHES.MODULE_ENTRY entry, then finally the new
+    // CACHES.PATH ENTRY.
+    realPathKey = genRealPathKey(realPath);
+
+    if (caches.MODULE_ENTRY.has(realPathKey)) {
+      caches.PATH.set(nameDirKey, realPathKey);
+      return caches.MODULE_ENTRY.get(realPathKey);
+    }
+
+    var thePackage = tryRequire(realPath);
```

Comment:
document why this is needed

Reply:
Sorry, which 'this' do you mean - tryRequire?

Reply:
yes `tryRequire`

Reply:
Added the following comment: 
    // require the package.  If we get a 'module-not-found' error it will return null
    // and we can insert an entry in the cache saying we couldn't find the package
    // in case it's requested later. Any other more serious error we specifically
    // don't try to catch here.
If you think this is unnecessary or overkill, I'm okay with just doing a require w/o tryRequire, just let me know. We might want to put out a console message too - thoughts?